AD shell element

sveerara opened the pull request:
Adds strain, stress, elasticity tensor and stressdivergence classes for quasi-static shell elements. Also adds documentation and tests for shell elements. 

refs #14280

sveerara committed:
AD shell element #14280

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/422258/) on 808aff1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14281/site)

This comment will be updated on new commits.

yw5aj commented on modules/tensor_mechanics/include/kernels/ADStressDivergenceShell.h:
```diff
@@ -0,0 +1,66 @@
+/**
+ * StressDivergenceTensors mostly copies from StressDivergence.  There are small changes to use
+ * RankFourTensor and RankTwoTensors instead of SymmElasticityTensors and SymmTensors.  This is done
+ * to allow for more mathematical transparancy.
```
Thank you Swetha for the nice PR! I do have a question - could you explain a bit on what you are referring to by "mathematical transparency"? I was thinking that for shells we are using the same symmetric elasticity & stress/strain tensors, but maybe you were referring to something else that I didn't think of.

sveerara replied:
Thanks for reviewing the PR, Shawn. This description actually belongs to StressDivergenceTensors class. I forgot to edit the description after making a copy of the file. Thanks for pointing this out. This comment does not apply for shells.

yw5aj replied:
Ah OK! Thanks Swetha for explaining.

sveerara commented:
@bwspenc, @acasagran -  please review this merge request when you find time. Thanks!

lindsayad commented on framework/include/utils/MooseTypes.h:
```diff
@@ -26,6 +26,7 @@
 #include "Eigen/Core"
 #include "libmesh/restore_warnings.h"
 #include "libmesh/tensor_tools.h"
+#include "libmesh/dense_matrix.h"
```
Instead of including this header, please just use a forward declaration:
```c++
namespace libMesh
{
template <typename>
class DenseMatrix;
}
```
since you are only doing `typedefs` here. You do not need the complete type

lindsayad commented on modules/tensor_mechanics/src/kernels/ADStressDivergenceShell.C:
```diff
@@ -0,0 +1,99 @@
+    ADStressDivergenceShell,
+    ADKernel,
+    params.addClassDescription("Quasi-static stress divergence kernel for Shell element");
+    params.addRequiredParam<unsigned int>("component",
```
Let's try to use `addRequiredRangeCheckedParam`

lindsayad commented on modules/tensor_mechanics/test/tests/shell/static/abaqus_plate_bending.i:
```diff
@@ -0,0 +1,207 @@
+  ymax = 20.0
+[]
+
+[MeshModifiers]
```
No `MeshModifiers` please. Please using the corresponding `MeshGenerator`.

lindsayad commented on modules/tensor_mechanics/test/tests/shell/static/tests:
```diff
@@ -0,0 +1,45 @@
+  [../]
+  [./large_beam_bending_jacobian]
+    type = 'PetscJacobianTester'
+    input = 'beam_bending_moment_AD.i'
```
It looks like you're just running the same input file as above without any `cli_args`, so it looks like an identical test. Probably not what you meant to do based on the different requirements?

lindsayad reviewed (changes requested):
Others can do a more meaty content review. Seems like a pretty neat feature!

sveerara committed:
Addressing comments #14281

sveerara commented:
@lindsayad, I have addressed your comments. Thanks!

lindsayad reviewed (dismissed):
Thanks for addressing comments @sveerara! I'm going to wait to merge for a physics person's review...

acasagran commented on modules/tensor_mechanics/test/tests/shell/static/plate_bending.i:
```diff
@@ -0,0 +1,199 @@
+# The FEM solution at (0,0), which is at the center of the full plate
+# is -2.997084e-03 m.
+
+# The analytical solution for displacement at center of plate obtained
```
Where did the analytical solution come from? Roark's or something else?

sveerara replied:
I have added the reference for the analytical solution.

acasagran commented on modules/tensor_mechanics/test/tests/shell/static/plate_bending.i:
```diff
@@ -0,0 +1,199 @@
+# m & n = 7).
+
+# The resulting relative error between FEM and analytical solution is
+# 0.048%.
```
None of the tests output stresses or strains. Can you add those just to see what they look like? Does it make sense for any of these tests to compare stress or strain to analytical solutions as well?

sveerara replied:
I have added the capability to compute the stresses/strains in the global coordinate system. This capability is tested in the beam bending test. 

Also, I have added a two other tests, a rotated cantilever beam with the beam in the XZ plane and a pinched cylinder problem to test the element in other coordinate systems.

dschwen commented on modules/tensor_mechanics/include/kernels/ADStressDivergenceShell.h:
```diff
@@ -0,0 +1,65 @@
+
+#include "ADKernel.h"
+#include "RankTwoTensor.h"
+#include "RankFourTensor.h"
```
you could forward declare those like you do further below

sveerara replied:
I have made this change

dschwen commented on modules/tensor_mechanics/include/materials/ADComputeFiniteShellStrain.h:
```diff
@@ -0,0 +1,40 @@
+declareADValidParams(ADComputeFiniteShellStrain);
+
+template <ComputeStage compute_stage>
+class ADComputeFiniteShellStrain : public ADComputeIncrementalShellStrain<compute_stage>
```
missing Doxygen comment

```
/**
 * ...
 */
template <...
```

dschwen commented on modules/tensor_mechanics/include/materials/ADComputeIsotropicElasticityTensorShell.h:
```diff
@@ -0,0 +1,54 @@
+
+#include "ADMaterial.h"
+
+#define usingComputeIsotropicElasticityTensorShellMembers usingMaterialMembers;
```
no semicolon here

dschwen commented on modules/tensor_mechanics/include/materials/ADComputeShellStress.h:
```diff
@@ -0,0 +1,56 @@
+#include "RankFourTensor.h"
+#include "ADComputeIsotropicElasticityTensorShell.h"
+
+#define usingComputeShellStressMembers usingMaterialMembers;
```
and here

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+!media media/tensor_mechanics/shell_representation.png
+      style=width:50%;margin-left:2%;float:right
+      id=fig_shell
+      caption=Shell element with 4 nodes and 3 translational and 2 rotational degrees of freedom at each node.
```
Looks like this image was pulled from that paper. We should at a minimum say it was taken from that paper, but to be safe, we probably shouldn't even include the image. I'd say for now, let's pull it out, and maybe make our own image.

sveerara replied:
I have added a different figure.

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+
+Each of the four nodes have 5 degrees of freedom - 3 displacements ($u_1^k$,$u_2^k$, $u_3^k$) and 2 rotations ($\alpha_k$ and $\beta_k$). The thickness ($a_k$) is assumed to be constant across the shell element and the initial normal ($^oV_n^k) at all nodes are same as the normal to the shell element, which is automatically computed using the coordinates of the nodes.
+
+The basic equation of motion for a quasi-static shell is the same as that of a continuum brick element:
```
Is it better to say 'equation of equilibrium'?

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+^t x_i =  \sum_{k=1}^4 h_k ^t {x_i}^k + \frac{r_3}{2} \sum_{k=1}^4 a_k h_k ^tV_{ni}^k
+\end{equation}
+
+Therefore, the incremental displacements of any point with natural coordinate $r_i$ of the shell element at time $t$ in the global cartesian coordinate system ($^ox_1$, $^ox_2$ and $^ox_3$) are:
```
'any point with natural coordinate $r_i$ of the shell element' -> 'any point within the shell element with natural coordinate $r_i$'

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+\tilde{\epsilon}_{ij} = \frac{1}{2}(^t \mathbf{g_i} \cdot ^t \mathbf{g_j} - ^0 \mathbf{g_i} \cdot ^0 \mathbf{g_j})
+\end{equation}
+
+The expressions for the small and large strains in the 11, 22, 12, 13 and 23 directions are provided in equations 21 to 24 of [!cite](dvorkin1984continuum). The expressions for the shear strains in the 13 and 23 directions also include a correction for shear locking. In Moose the small strain increments are computed in
```
Moose->MOOSE

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+\end{equation}
+
+The expressions for the small and large strains in the 11, 22, 12, 13 and 23 directions are provided in equations 21 to 24 of [!cite](dvorkin1984continuum). The expressions for the shear strains in the 13 and 23 directions also include a correction for shear locking. In Moose the small strain increments are computed in
+[ADComputeIncrementalShellStrain](/ADComputeIncrementalShellStrain.md) and the small + large strain increments are computed in [ADComputeFiniteShellStrain](/ADComputeFiniteShellStrain.md)). Note that strain in the 33 direction is not computed and is set to zero.
```
Is that right about the strain in the 33 direction being 0?  I thought the stress was 0 and we computed the strain to give that.

sveerara replied:
I have corrected this statement. We are not computing the strain in the 33 direction. Instead the elasticity tensor is condensed appropriately.

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+
+Finally the residuals are computed and assembled in [ADStressDivergenceShell](/ADStressDivergenceShell.md) for both small and large strain scenarios. For small strain scenarios, the residual at each qp is computed by multiplying the stress tensor at that qp with the corresponding components of the B matrix. Additionally, for large strain scenarios, the old stress tensor is multiplied with corresp ...
+
+Note that there are qps both along the plane as well as along the thickness of the element. The order of Gauss quadrature rule along the thickness is provided as input by the user to all the above mentioned objects.  
```
qps -> quadrature points

bwspenc commented on modules/tensor_mechanics/doc/content/source/kernels/ADStressDivergenceShell.md:
```diff
@@ -0,0 +1,15 @@
+
+## Description
+
+The `ADStressDivergenceShell` kernel calculates the residual of the stress divergence for shell elements. Forward mode automatic differentiation is used to compute an exact Jacobian. Please refer to [ShellElements](/ShellElements.md) for details.
```
the residual of the -> the contribution to the residual from the

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADComputeIncrementalShellStrain.md:
```diff
@@ -0,0 +1,13 @@
+
+## Description
+
+This class computes the small strain increments for shell elements. Please refer to [ShellElements](/ShellElements.md) for details.
```
small strain increments -> incremental small strain

bwspenc commented on modules/tensor_mechanics/include/kernels/ADStressDivergenceShell.h:
```diff
@@ -0,0 +1,65 @@
+  std::vector<const ADMaterialProperty(RankTwoTensor) *> _stress;
+  std::vector<const MaterialProperty<RankTwoTensor> *> _stress_old;
+  std::vector<const ADMaterialProperty(DenseMatrix<Real>) *> _B_mat;
+  std::vector<const ADMaterialProperty(DenseMatrix<Real>) *> _B_NL;
```
I'd re-name _B_NL -> _B_nl, and _Jmap-> _J_map to be more consistent with MOOSE naming conventions.

bwspenc commented on modules/tensor_mechanics/include/kernels/ADStressDivergenceShell.h:
```diff
@@ -0,0 +1,65 @@
+  /// Quadrature rule in the out of plane direction
+  std::unique_ptr<QGauss> _t_qrule;
+
+  /// Quadrature points in the out of plane direction in isoparametric coordinate system
```
It looks like these are weights rather than points

bwspenc commented on modules/tensor_mechanics/include/materials/ADComputeIncrementalShellStrain.h:
```diff
@@ -0,0 +1,210 @@
+protected:
+  virtual void initQpStatefulProperties() override;
+  virtual void computeProperties() override;
+  virtual void computeSolnVector();
```
Need doxygen comments for non-override methods

bwspenc commented on modules/tensor_mechanics/include/materials/ADComputeIncrementalShellStrain.h:
```diff
@@ -0,0 +1,210 @@
+  /// Vector that stores the incremental solution at all the 20 DOFs in the 4 noded element.
+  ADDenseVector _soln_vector;
+
+  ADDenseVector _strain_vector;
```
doxygen

bwspenc commented on modules/tensor_mechanics/include/materials/ADComputeIncrementalShellStrain.h:
```diff
@@ -0,0 +1,210 @@
+  std::vector<const MaterialProperty<RealVectorValue> *> _dxyz_dzeta_old;
+
+  /// First tangential vectors at nodes
+  std::vector<ADRealVectorValue> _V1;
```
I'd prefer to keep with the MOOSE naming convention and keep variables all lower case. Sometimes, we make exceptions for matrices. Seems like these _V1 and _V2 vectors could be _v1 and _v2

bwspenc commented on modules/tensor_mechanics/include/materials/ADComputeIncrementalShellStrain.h:
```diff
@@ -0,0 +1,210 @@
+  /// Old ge matrix for elasticity tensor conversion
+  std::vector<const MaterialProperty<RankTwoTensor> *> _ge_old;
+
+  /// Material property containing jacobian of transformation
```
_J_map would be more consistent

bwspenc commented on modules/tensor_mechanics/include/materials/ADComputeIncrementalShellStrain.h:
```diff
@@ -0,0 +1,210 @@
+  ADRealVectorValue _x3;
+  const NumericVector<Number> * const & _sol;
+  const NumericVector<Number> & _sol_old;
+  ADRealVectorValue _g3_A;
```
Consider making these lowercase

bwspenc commented on modules/tensor_mechanics/src/kernels/ADStressDivergenceShell.C:
```diff
@@ -0,0 +1,101 @@
+    params.addRequiredRangeCheckedParam<unsigned int>(
+        "component",
+        "component < 5",
+        "An integer corresponding to the direction "
```
direction -> degree of freedom
this kernel acts in -> this kernel acts on

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeFiniteShellStrain.C:
```diff
@@ -0,0 +1,378 @@
+
+  unsigned int dim = _current_elem->dim();
+  if ((dim != 2))
+    mooseError("Shell element is implemented only for 2D Linear elements");
```
Linear -> linear

bwspenc replied:
Can't we do this check earlier on so we're not doing it every material evaluation?  Also, probably need to check for the order of the element and for the fact that it's restricted to 4-noded elements as well (it doesn't work for triangles, right?)

sveerara replied:
I have moved this check to initQpStatefulProperties of ADComputeIncrementalShellStrain. 

I have also added a test to check if that the element is first order (in the constructor) and to check if the element has 4 nodes (in initQpStatefulProperties).

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+        "The displacements appropriate for the simulation geometry and coordinate system");
+    params.addRequiredCoupledVar(
+        "thickness",
+        "Cross-section area of the beam. Can be supplied as either a number or a variable name.");
```
beam ->shell

bwspenc replied:
Cross-section area -> thickness

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeFiniteShellStrain.C:
```diff
@@ -0,0 +1,378 @@
+  {
+    for (unsigned int j = 0; j < _t_points.size(); ++j)
+    {
+      (*_BNL[j])[i].resize(5, 20);
```
I'd prefer that _BNL be named _B_nl to be more consistent

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+        &getMaterialPropertyOldByName<RankTwoTensor>("total_strain_t_points_" + std::to_string(i));
+    _B[i] = &declareADProperty<DenseMatrix<Real>>("B_matrix_t_points_" + std::to_string(i));
+    _B_old[i] =
+        &getMaterialPropertyOldByName<DenseMatrix<Real>>("B_matrix_t_points_" + std::to_string(i));
```
Maybe no need to include _matrix in the name

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+
+  unsigned int dim = _current_elem->dim();
+  if ((dim != 2))
+    mooseError("Shell element is implemented only for 2D Linear elements");
```
Same issues I mentioned before about this check -- do it once upfront if possible, and check for more things

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+
+  unsigned int dim = _current_elem->dim();
+  if ((dim != 2))
+    mooseError("Shell element is implemented only for 2D Linear elements");
```
No need to re-do this check here

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+    (*_dxyz_dzeta[t])[_qp] = c;
+  }
+
+  // calculating derivatives of shape function is physical space (dphi/dx, dphi/dy, dphi/dz) at
```
is -> in

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+ADComputeIncrementalShellStrain<compute_stage>::computeBMatrix()
+{
+  // compute nodal local axis
+  _x2(1) = 1;
```
We could do this once up-front in the constructor

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -0,0 +1,544 @@
+        (*_B[j])[i](3, 2 + component * 4) = 1.0 / 8.0 * (1.0 + _2d_points[i](1)) * _g3_A(component);
+        (*_B[j])[i](3, 3 + component * 4) =
+            1.0 / 8.0 * (1.0 + _2d_points[i](1)) * -_g3_A(component);
+        (*_B[j])[i](3, 1 + component * 4) = 1.0 / 8.0 * (1.0 - _2d_points[i](1)) * _g3_C(component);
```
replace the instances of 1/8 with 0.125.  Less operations and still self-explanatory.

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeIsotropicElasticityTensorShell.C:
```diff
@@ -0,0 +1,88 @@
+defineADValidParams(
+    ADComputeIsotropicElasticityTensorShell,
+    ADMaterial,
+    params.addClassDescription("Compute a plane stress isotropic elasticity tensor.");
```
Just a comment -- we could probably use this for plane stress elasticity problems as well, except the fact that we have the through-thickness index requires a little more work.

bwspenc commented on modules/tensor_mechanics/src/materials/ADComputeShellStress.C:
```diff
@@ -0,0 +1,72 @@
+defineADValidParams(
+    ADComputeShellStress,
+    ADMaterial,
+    params.addClassDescription("Compute stress using elasticity for shell");
```
This is the stress in the in-plane coordinate system, right?

bwspenc commented on modules/tensor_mechanics/test/tests/shell/static/abaqus_plate_bending.i:
```diff
@@ -0,0 +1,208 @@
+# Shell element verification test from Abaqus verification manual 1.3.13
```
Does this happen to be a NAFEMS problem?  Many of them are. I'd rather reference the NAFEMS benchmark than another code's verification manual if possible.

sveerara replied:
I currently don't have access to the NAFEMS benchmark problems. So, I don't know whether it is a NAFEMS problem.

bwspenc replied:
Maybe it's not.  Could we still keep the comment there pointing to the abaqus problem, but rename the input file to something other than abaqus_plate_bending? Maybe just plate_bending?

bwspenc commented on modules/tensor_mechanics/test/tests/shell/static/abaqus_plate_bending.i:
```diff
@@ -0,0 +1,208 @@
+
+[NodalKernels]
+  [./force_C]
+    type = ConstantRate
```
Do we really use an object called ConstantRate to apply nodal forces?!  That's a super obtuse name!

sveerara replied:
Yes, this is an interesting name :)

bwspenc commented on modules/tensor_mechanics/test/tests/shell/static/abaqus_plate_bending.i:
```diff
@@ -0,0 +1,208 @@
+    youngs_modulus = 1e3
+    poissons_ratio = 0.3
+    block = 0
+    order = SECOND
```
Looking at this input, it seems to me like we need a different name to define the integration order for the out-of-plane integration to avoid confusion. Maybe 'thickness_order'? 'through_thickness_order'?

sveerara replied:
I have changed 'order' to 'through_thickness_order'.

permcody commented:
@bwspenc - Not that she's gone, this will need to be picked up by another developer or it will get stale botted!

bwspenc commented:
Yeah...

sveerara commented:
@bwspenc and @permcody, I worked on some of the changes and have a few more to go. I will push the updated version of the shell element in the next few days. Thanks!

sveerara committed:
Addressing second round of comments #14281

sveerara commented:
@acasagran, @dschwen and @bwspenc, I have addressed your comments. Please let me know any additional changes need to be made. Thanks!

Happy new year :)

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/ShellElements.md:
```diff
@@ -0,0 +1,78 @@
+
+## Stress divergence
+
+Finally the residuals are computed and assembled in [ADStressDivergenceShell](/ADStressDivergenceShell.md) for both small and large strain scenarios. For small strain scenarios, the residual at each qp is computed by multiplying the stress tensor at that qp with the corresponding components of the B matrix. Additionally, for large strain scenarios, the old stress tensor is multiplied with corresp ...
```
Never mind -- I see that you introduced the qp acronym earlier

moosebuild commented:
Job [Test debug](https://civet.inl.gov/job/452949/) on f754658 : invalidated by @bwspenc

Failure unrelated to this PR

permcody commented:
@bwspenc - There is a small diff in one of the new parallel tests added in this PR. Do you or someone on your team have time to look at it? If not, we can force it to serial for now. The diff looks small on values that are close to zero (normal relative diff issues we usually see in mechanics).

bwspenc commented:
@permcody Whoops! Thanks for letting me know. @dylanm is looking into it.

bwspenc commented:
Actually, @dylanjm
